validator/client/src/main/java/tech/pegasys/teku/validator/client/DoppelgangerDetection.java

```diff
@@ -0,0 +1,288 @@
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.validator.api.ValidatorApiChannel;
+import tech.pegasys.teku.validator.beaconnode.GenesisDataProvider;
+
+public class DoppelgangerDetection {
+
+  private static final Logger LOGGER = LogManager.getLogger();
+  private final Duration checkDelay;
+  private final Duration timeout;
+  private final AsyncRunner asyncRunner;
+  private final ValidatorApiChannel validatorApiChannel;
+  private final ValidatorIndexProvider validatorIndexProvider;
+  private final Spec spec;
+  private final TimeProvider timeProvider;
+  private final GenesisDataProvider genesisDataProvider;
+  private Optional<UInt64> epochAtStart = Optional.empty();
+  private volatile Cancellable doppelgangerDetectionTask;
+  private Optional<SafeFuture<Map<Integer, BLSPublicKey>>> doppelgangerCheckFinished =
+      Optional.empty();
+  private long startTime;
+
+  public DoppelgangerDetection(
+      final AsyncRunner asyncRunner,
+      final ValidatorApiChannel validatorApiChannel,
+      final ValidatorIndexProvider validatorIndexProvider,
+      final Spec spec,
+      final TimeProvider timeProvider,
+      final GenesisDataProvider genesisDataProvider,
+      final Duration checkDelay,
+      final Duration timeout) {
+    this.asyncRunner = asyncRunner;
+    this.validatorApiChannel = validatorApiChannel;
+    this.validatorIndexProvider = validatorIndexProvider;
+    this.spec = spec;
+    this.timeProvider = timeProvider;
+    this.genesisDataProvider = genesisDataProvider;
+    this.checkDelay = checkDelay;
+    this.timeout = timeout;
+  }
+
+  protected synchronized SafeFuture<Map<Integer, BLSPublicKey>> performDoppelgangerDetection() {
+    LOGGER.info("Starting doppelganger detection...");
+    doppelgangerCheckFinished = Optional.of(new SafeFuture<>());
+    startTime = System.nanoTime();
+    doppelgangerDetectionTask =
+        asyncRunner.runWithFixedDelay(
+            this::checkValidatorsDoppelganger,
+            checkDelay,
+            checkDelay,
+            throwable ->
+                LOGGER.error(
+                    "Error while checking validators doppelganger. The check could not be performed correctly.",
+                    throwable));
+    return doppelgangerCheckFinished.get();
+  }
+
+  private synchronized SafeFuture<?> stopDoppelgangerDetection(
+      Map<Integer, BLSPublicKey> detectedDoppelgangers) {
+    if (doppelgangerCheckFinished.isEmpty()) {
+      throw new RuntimeException("Doppelganger Detection is already stopped.");
+    }
+    doppelgangerCheckFinished.get().complete(detectedDoppelgangers);
+    doppelgangerCheckFinished = Optional.empty();
+    return SafeFuture.fromRunnable(
+        () -> Optional.ofNullable(doppelgangerDetectionTask).ifPresent(Cancellable::cancel));
+  }
+
+  private SafeFuture<?> checkValidatorsDoppelganger() {
+    Duration duration = Duration.of(System.nanoTime() - startTime, ChronoUnit.NANOS);
```

Comment:
You can use `SafeFuture<Void>` hear and in stopDoppelgangerDetection  instead of `SafeFuture<?>` (it isn't a service anymore). 
so you can use `toVoid()` instead of `.thenApply(o -> null);` down there.